### PR TITLE
Change DistributedContextSdk.Builder to use current dctx as implicit parent

### LINKED ISSUE #436: SDK: DistributedContextSdk.Builder should use current DistributedContext as implicit parent
see https://github.com/open-telemetry/opentelemetry-java/pull/428#discussion_r297252224

### PR DESCRIPTION
**Pull request by arminru**
Fixes #436

API definition:
https://github.com/open-telemetry/opentelemetry-java/blob/ed42237b6c71052bbffe4d849647c53fae81b246/api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java#L57-L61

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/distributedcontext/DistributedContextSdk.java [hunk 1]
```diff
@@ -151,7 +152,9 @@ public int hashCode() {
 
     @Override
     public DistributedContextSdk build() {
-      // TODO if (parent == null) parent = DistributedContextManager.getCurrentContext();
+      if (parent == null) {
```
Comment (songy23) [inline sdk/src/main/java/io/opentelemetry/sdk/distributedcontext/DistributedContextSdk.java]: I've looked into this approach too - the problem with this is if a user called `setNoParent()`, that will set `parent` to null, but if at the same time there's a distributed context in the current context, `build()` will still return a child distributed context. I don't think this is what users wanted:
```java
try (Scope scope = contextManager.withContext(root)) {
  Builder builder = DistributedContextSdk.builder().setNoParent(); // I want another root.
  builder.build(); // this will give me a child. 
}
```

Comment (arminru) [reply sdk/src/main/java/io/opentelemetry/sdk/distributedcontext/DistributedContextSdk.java]: I implemented it according to the spec as defined here:
https://github.com/open-telemetry/opentelemetry-java/blob/ed42237b6c71052bbffe4d849647c53fae81b246/api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java#L76-L84

I already found the spec to be contradictory (setNoParent --> will use X as parent) and was wondering why `setNoParent` was there in first place then (if someone wouldn't want an explicit parent why would they call `setParent` at all just to call `setNoParent` afterwards?) but was assured it's correct and should stay there.

Looks like the second sentence should rather be removed from the spec, right?
Then I'd introduce a flag which is set using `setNoParent` and would prevent getting the implicit parent at `build()`.

File: api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java [hunk 2]
```diff
@@ -74,9 +74,7 @@
     Builder setParent(DistributedContext parent);
 
     /**
-     * Sets the option to become a {@link DistributedContext} with no parent. If not set, the value
-     * of {@link DistributedContextManager#getCurrentContext()} at {@link #build()} time will be
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java]: I think this statement still applies. If `setNoParent` is never called, implicit parent will still be used, while if `setNoParent` is called, neither implicit nor explicit parent will be used.

Comment (arminru) [reply api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java]: I'll rephrase it to clarify it a bit better.

File: api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java [hunk 3]
```diff
@@ -56,9 +56,9 @@
    */
   interface Builder {
     /**
-     * Sets the parent {@link DistributedContext} to use. If not set, the value of {@link
-     * DistributedContextManager#getCurrentContext()} at {@link #build()} time will be used as
-     * parent.
+     * Sets the parent {@link DistributedContext} to use. If no parent (or {@code null}) is
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java]: Not sure it is clear to me what the behavior should be if null. Probably we should not allow null.

Comment (arminru) [reply api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java]: With the builder pattern something like "setParent(getMyContext()) but only if it's not null" is a bit cumbersome to write but on the other hand it could be surprising that an implicit parent is used after the user tried setting another one which turned out to be null, yes.
I'm fine either way. What would you prefer, @songy23?

Comment (songy23) [reply api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java]: IMO we can rephrase the statement here to be "If you want to set an explicit parent, use setParent(). Explicit parent cannot be null. If you want to use the value in current context (which could be null), you don't need to call this method." Then we can enforce `setParent` to take non-null.

Comment (arminru) [reply api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java]: I missed the `@throws NullPointerException if {@code parent} is {@code null}` declaration we already have in setParent's doc. The current `DistributedContextSdkTest.setParent_nullValue` also asserted a different behavior (clearing previously set parent).
I think this leads to least surprise (no implicit parent if setParent is called with null) so I'll implement it for now - this also assures consistency with `NoopDistributedContextBuilder`. Once https://github.com/open-telemetry/opentelemetry-specification/pull/153 is settled this can be revisited.

**Comment by bogdandrutu** [comment]
Please rebase.
